syntax = "proto3";

package yandex.cloud.events.iam;

import "google/rpc/status.proto";
import "yandex/cloud/access/access.proto";
import "yandex/cloud/events/common.proto";
import "yandex/cloud/validation.proto";
import "yandex/cloud/events/options.proto";

option go_package = "a.yandex-team.ru/cloud/bitbucket/public-api/yandex/cloud/events/iam;iam";
option java_package = "yandex.cloud.api.events.iam";
option java_outer_classname = "ServiceAccountEvents";

message CreateServiceAccount {
  option (include) = true;

  Authentication authentication = 1 [(required) = true];
  Authorization authorization = 2 [(required) = true];
  EventMetadata event_metadata = 3 [(required) = true];
  RequestMetadata request_metadata = 4 [(required) = true];
  EventStatus event_status = 5 [(required) = true];
  google.rpc.Status error = 6;
  EventDetails details = 7 [(required) = true];

  message EventDetails {
    string service_account_id = 1 [(required) = true];
    string service_account_name = 2 [(required) = true];
  }

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    ERROR = 2;
    DONE = 3;
  }
}

message UpdateServiceAccount {
  option (include) = true;

  Authentication authentication = 1 [(required) = true];
  Authorization authorization = 2 [(required) = true];
  EventMetadata event_metadata = 3 [(required) = true];
  RequestMetadata request_metadata = 4 [(required) = true];
  EventStatus event_status = 5 [(required) = true];
  google.rpc.Status error = 6;
  EventDetails details = 7 [(required) = true];

  message EventDetails {
    string service_account_id = 1 [(required) = true];
    string service_account_name = 2 [(required) = true];
  }

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    ERROR = 2;
    DONE = 3;
  }
}

message DeleteServiceAccount {
  option (include) = true;

  Authentication authentication = 1 [(required) = true];
  Authorization authorization = 2 [(required) = true];
  EventMetadata event_metadata = 3 [(required) = true];
  RequestMetadata request_metadata = 4 [(required) = true];
  EventStatus event_status = 5 [(required) = true];
  google.rpc.Status error = 6;
  EventDetails details = 7 [(required) = true];

  message EventDetails {
    string service_account_id = 1 [(required) = true];
    string service_account_name = 2 [(required) = true];
  }

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    STARTED = 1;
    ERROR = 2;
    DONE = 3;
  }
}

message UpdateServiceAccountAccessBindings {
  option (include) = true;

  Authentication authentication = 1 [(required) = true];
  Authorization authorization = 2 [(required) = true];
  EventMetadata event_metadata = 3 [(required) = true];
  RequestMetadata request_metadata = 4 [(required) = true];
  EventStatus event_status = 5 [(required) = true];
  google.rpc.Status error = 6;
  EventDetails details = 7 [(required) = true];

  message EventDetails {
    string service_account_id = 1 [(required) = true];
    string service_account_name = 2 [(required) = true];
    repeated access.AccessBindingDelta access_binding_deltas = 3 [(required) = true];
  }

  enum EventStatus {
    EVENT_STATUS_UNSPECIFIED = 0;
    STARTED = 1;
    ERROR = 2;
    DONE = 3;
  }
}
